File: MatlabOctaveCodeSnippets/snip_systems_IIR_design.m
Wp=0.3; %normalized passband frequency (rad / pi)
Ws=0.7; %normalized stopband frequency (rad / pi)
Rp=1; %maximum passband ripple in dB
Rs=40; %minimum stopband attenuation in dB
[n,Wn] = buttord(Wp,Ws,Rp,Rs); %find the Butterworth order
[B,A]=butter(n,Wn); %design the Butterworth filter
zp=exp(1j*pi*Wp); zs=exp(1j*pi*Ws); %Z values at Wp and Ws
%% Check if the filter obeys requirements
linearMagWp = abs(polyval(B,zp)/polyval(A,zp)) %mag. at Wp
linearMagWs = abs(polyval(B,zs)/polyval(A,zs)) %mag. at Ws
Dp_result = 1-linearMagWp %linear deviation at Wp (linear)
Rp_result = -20*log10(1-Dp_result) %deviation (ripple) at Wp (dB)
Ds_result = linearMagWs %linear deviation at Ws (linear)
Rs_result = -20*log10(Ds_result) %deviation (ripple) at Ws (dB)

File: PythonCodeSnippets/snip_systems_IIR_design.py
import numpy as np
from scipy.signal import buttord, butter

# Normalized passband frequency (rad / pi)
Wp = 0.3
# Normalized stopband frequency (rad / pi)
Ws = 0.7
# Maximum passband ripple in dB
Rp = 1
# Minimum stopband attenuation in dB
Rs = 40

# Find the Butterworth order
n, Wn = buttord(Wp, Ws, Rp, Rs)
# Design the Butterworth filter
B, A = butter(3 * n, Wn)

# Z values at Wp and Ws
zp = np.exp(1j * np.pi * Wp)
zs = np.exp(1j * np.pi * Ws)

# Magnitude at Wp
linearMagWp = np.abs(np.polyval(B, zp) / np.polyval(A, zp))
# Magnitude at Ws
linearMagWs = np.abs(np.polyval(B, zs) / np.polyval(A, zs))

# Linear deviation at Wp (linear)
D_p = 1 - linearMagWp
# Deviation (ripple) at Wp (dB)
R_p = -20 * np.log10(1 - D_p)
# Linear deviation at Ws (linear)
D_s = linearMagWs
# Deviation (ripple) at Ws (dB)
R_s = -20 * np.log10(D_s)

print("Linear deviation at Wp (linear):", D_p)
print("Deviation (ripple) at Wp (dB):", R_p)
print("Linear deviation at Ws (linear):", D_s)
print("Deviation (ripple) at Ws (dB):", R_s)
